codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -327,6 +353,21 @@ static void fireChannelRead(ChannelHandlerContext ctx, CodecOutputList msgs, int
         }
     }
 
+    /**
+     * Get {@code numElements} out of the {@link CodecOutputList} and batch forward these through the pipeline.
+     */
+    static void batchFireChannelRead(ChannelHandlerContext ctx, CodecOutputList msgs, int numElements) {
+        if (numElements == 1) {
+            ctx.fireChannelRead(msgs.get(0));
+        } else {
+            ArrayList<Object> ret = new ArrayList<Object>(numElements);
+            for (int i = 0; i < numElements; i++) {
+                ret.add(msgs.get(i));
```

Comment:
`List<Object> ret = new ArrayList<Object>(msgs);` will be better.

Reply:
might consider using RecyclableArrayList as the majority of pipeline code is essentialy trash-free
personally I've had issues with using RecyclableArrayList, it was causing weird bugs...but after endless digging I gave up and returned my code to pure ArrayList...
or maybe we can re-use one ArrayList only(clear after fireChannelRead)...clients don't need to access the passed ArrayList outside of pipeline...

Reply:
also, we could reduce the interface used down to Collection or something, in case a more efficient class is to be used later?